### PR TITLE
GroupBy: Cap dictionary-building selector memory usage.

### PR DESCRIPTION
**Pull request by gianm**
New context parameter "maxSelectorDictionarySize" controls when the
per-segment processing code should return early and trigger a trip
to the merge buffer.

Includes:

- Vectorized and nonvectorized implementations.
- Adjustments to GroupByQueryRunnerTest to exercise this code in
  the v2SmallDictionary suite. (Both the selector dictionary and
  the merging dictionary will be small in that suite.)
- Tests for the new config parameter.

### CODE REVIEW

**Review by clintropolis — APPROVED** [review]
:+1:

File: processing/src/main/java/org/apache/druid/query/groupby/epinephelinae/DictionaryBuilding.java [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Utilities for parts of the groupBy engine that need to build dictionaries.
+ */
+public class DictionaryBuilding
+{
+  public static final int UNKNOWN_DICTIONARY_ID = -1;
```
[the remark this answers is not in the document]
Comment (gianm) [reply processing/src/main/java/org/apache/druid/query/groupby/epinephelinae/DictionaryBuilding.java:33]: I switched to DimensionDictionary.ABSENT_VALUE_ID.

File: processing/src/main/java/org/apache/druid/query/groupby/epinephelinae/column/DictionaryBuildingStringGroupByColumnSelectorStrategy.java [hunk 2]
```diff
@@ -44,13 +44,11 @@
 {
   private static final int GROUP_BY_MISSING_VALUE = -1;
 
-  private int nextId = 0;
-  private final List<String> dictionary = new ArrayList<>();
-  private final Object2IntOpenHashMap<String> reverseDictionary = new Object2IntOpenHashMap<>();
+  // Entry in dictionary, node pointer in reverseDictionary, hash + k/v/next pointer in reverseDictionary nodes
+  public static final int ROUGH_OVERHEAD_PER_DICTIONARY_ENTRY = Long.BYTES * 5 + Integer.BYTES;
```
[the remark this answers is not in the document]
Comment (gianm) [reply processing/src/main/java/org/apache/druid/query/groupby/epinephelinae/column/DictionaryBuildingStringGroupByColumnSelectorStrategy.java:48]: yep, it is. Fixed.

**Review by clintropolis — APPROVED** [review]

**Review by rohangarg — APPROVED** [review]
LGTM!

**Pull request merged**